e2e/nomad09upgrade/upgrade.go

```diff
@@ -0,0 +1,339 @@
+	os.RemoveAll(tc.binDir)
+}
+
+func (tc *UpgradePathTC) TestRawExecTaskUpgrade(f *framework.F) {
+	for _, ver := range nomadVersions {
+		ver := ver
+		f.T().Run(ver, func(t *testing.T) {
+			t.Parallel()
+			tc.testUpgradeForJob(t, ver, "rawexec.nomad")
+		})
+	}
+}
+
+func (tc *UpgradePathTC) TestExecTaskUpgrade(f *framework.F) {
+	for _, ver := range nomadVersions {
+		ver := ver
+		f.T().Run(ver, func(t *testing.T) {
+			t.Parallel()
+			tc.testUpgradeForJob(t, ver, "exec.nomad")
+		})
+	}
+}
+
+func (tc *UpgradePathTC) testUpgradeForJob(t *testing.T, ver string, jobfile string) {
+	require := require.New(t)
+	// Start a nomad agent for the given version
+	srv, err := tc.newNomadServer(t, ver)
+	require.NoError(err)
+	t.Logf("launching v%s nomad agent", ver)
+	require.NoError(srv.StartAgent())
+
+	// Wait for the agent to be ready
+	client, err := srv.Nomad()
+	require.NoError(err)
+	time.Sleep(time.Second * 5)
+	e2eutil.WaitForNodesReady(t, client, 1)
+
+	// Register a sleep job
+	jobID := "sleep-" + uuid.Generate()[:8]
+	t.Logf("registering exec job with id %s", jobID)
+	e2eutil.RegisterAndWaitForAllocs(t, client, jobfile, jobID)
+	allocs, _, err := client.Jobs().Allocations(jobID, false, nil)
+	require.NoError(err)
+	require.Len(allocs, 1)
+
+	// Wait for sleep job to transistion to running
+	id := allocs[0].ID
+	e2eutil.WaitForAllocRunning(t, client, id)
+
+	// Stop the agent, leaving the sleep job running
+	require.NoError(srv.StopAgent())
+
+	// Start a nomad agent with the to be tested nomad binary
+	t.Logf("launching test nomad agent")
+	require.NoError(srv.StartTargetAgent())
+
+	// Wait for the agent to be ready
+	client, err = srv.Nomad()
+	require.NoError(err)
+	time.Sleep(time.Second * 5)
+	e2eutil.WaitForNodesReady(t, client, 1)
+
+	// Make sure the same allocation still exists
+	alloc, _, err := client.Allocations().Info(id, nil)
+	require.NoError(err)
+	// Pull stats from the allocation, testing that new code can interface with
+	// the old stats driver apis
+	testutil.WaitForResult(func() (bool, error) {
+		stats, err := client.Allocations().Stats(alloc, nil)
+		if err != nil {
+			return false, err
+		}
+
+		return stats.ResourceUsage.MemoryStats.RSS > 0, fmt.Errorf("RSS for task should be greater than 0")
+	}, func(err error) {
+		require.NoError(err)
+	})
+
+	// Deregister the job. This tests that the new code can properly tear down
+	// upgraded allocs
+	_, _, err = client.Jobs().Deregister(alloc.JobID, true, nil)
+	require.NoError(err)
+	testutil.WaitForResult(func() (bool, error) {
+		j, _, _ := client.Jobs().Info(jobID, nil)
+		if j == nil {
+			return true, nil
+		}
+		time.Sleep(time.Millisecond * 100)
+		return false, fmt.Errorf("job with id %q should be purged", jobID)
+	}, func(err error) {
+		require.NoError(err)
+	})
+
+	// Check that the task dir mounts have been removed
+	testutil.WaitForResult(func() (bool, error) {
+		defer client.System().GarbageCollect()
```

Comment:
Doing a GC on every loop seems excessive. Could just defer it to run when the test completes right?

Reply:
I'm not sure the best way to do this. I essentially need to wait for the alloc to be GCed and if the client hasn't marked it for GC then calling garbage collect will not remove it. So I'm essentially checking if mounts got cleaned up and GCing if not in a loop.